### DISCUSSION TITLE
how can i send print commands one line at a time with ImmediateNetworkPrinter?

### DISCUSSION THREAD

**Discussion by xZeroKooLx** [lukevp/ESC-POS-.NET#224] [Q&A] [answered]
I'm trying to use ImmediateNetworkPrinter but its jumbling up the order in witch the lines are sent to the printer and cutting before printing has finished. 
(please see attached image) The top 2 recepts are what i expect and get with NetworkPrinter but the bottom 2 are using ImmediateNetworkPrinter
![PXL_20230215_160756346 MP](https://user-images.githubusercontent.com/26562404/219084971-bb34986f-e7d1-4464-a643-9f504e017f52.jpg)

            
            if (printer == null) printer = new ImmediateNetworkPrinter(settings: new ImmediateNetworkPrinterSettings() { ConnectionString = $"{targetAddress}:{targetPort}" });
            Bitmap bmp = new Bitmap(listBox1.Width, listBox1.Height);
            listBox1.DrawToBitmap(bmp, new Rectangle(0, 0, bmp.Width, bmp.Height));

            e = new EPSON();
            // Convert the bitmap to a byte array     
            using (var ms = new MemoryStream())
            {
                bmp.Save(ms, ImageFormat.Png);
                byte[] img = ms.ToArray();
                // Print the image
                printer.WriteAsync(e.CenterAlign());
                printer.WriteAsync(e.PrintLine(customMsgRTextBox.Text));
                printer.WriteAsync(e.FeedLines(1));
                printer.WriteAsync(e.SetStyles(PrintStyle.Bold | PrintStyle.FontB | PrintStyle.DoubleHeight | PrintStyle.DoubleWidth));
                printer.WriteAsync(e.PrintLine("YOUR LUCKY NUMBERS ARE:"));
                printer.WriteAsync(e.FeedLines(1));
                printer.WriteAsync(e.PrintImage(img, true, true, 350, 1100));
                printer.WriteAsync(e.SetStyles(PrintStyle.None));
                printer.WriteAsync(e.FeedLines(1));
                printer.WriteAsync(e.PrintLine("Dont forget to put your name on the back of"));
                printer.WriteAsync(e.PrintLine("your ticket and place it in the red bin located"));
                printer.WriteAsync(e.PrintLine("in the bar for your chance to win 1 of 2x $100"));
                printer.WriteAsync(e.PrintLine("Telegraph vouvhers in our"));
                printer.WriteAsync(e.PrintLine("end of month draw!"));
                printer.WriteAsync(e.FeedLines(1));
                printer.WriteAsync(e.PrintLine("Thanks for visiting The Telegraph Gunning."));
                printer.WriteAsync(e.PrintLine(DateTime.Now.ToString("dd/MM/yy hh:mm:ss tt")));
                printer.WriteAsync(e.FeedLines(1));
                printer.WriteAsync(e.FullCutAfterFeed(1));
                Random rnd = new Random();
                int randomNumber = rnd.Next(1, 5); // generates a number between 1 and 10
                if (randomNumber == 1)
                {
                    //tndTicket();
                }
            }
`

**Comment by lukevp** [comment, marked as answer]
You have to await asynchronous calls in order to make sure they are complete before the next call. The standard in C# is to name methods with the `Async` suffix to indicate that they are asynchronous. Your IDE should have given you a warning as well. You can tell a call is Async also because it will return a Task rather than a direct value.

the reason your writes are out of order is that you’re not awaiting the methods properly. All of them are going in at once and they are competing in an indeterminate order.

 It works with the plain network printer because it internally buffers all writes and flushes them in order for you.

please give it a try with async/await and let me know if you still have an issue!

**Reply by xZeroKooLx** [reply]
Hi luke and thanks for the quick reply, I think I've modified my code to reflect your suggested changes but am still getting the same result.

        private async Task PrintNumbersAsync()
        {
            string filePath = "textfile.txt";
            string textFromFile = File.ReadAllText(filePath);
            string[] parts = textFromFile.Split(':');
            string targetAddress = parts[0];
            string targetPort = parts[1];

            if (printer == null) printer = new ImmediateNetworkPrinter(settings: new ImmediateNetworkPrinterSettings() { ConnectionString = $"{targetAddress}:{targetPort}" });

            Bitmap bmp = new Bitmap(listBox1.Width, listBox1.Height);
            listBox1.DrawToBitmap(bmp, new Rectangle(0, 0, bmp.Width, bmp.Height));

            e = new EPSON();
            using (var ms = new MemoryStream())
            {
                bmp.Save(ms, ImageFormat.Png);
                byte[] img = ms.ToArray();

                // Print the image
                await printer.WriteAsync(e.CenterAlign());
                await printer.WriteAsync(e.PrintLine(customMsgRTextBox.Text));
                await printer.WriteAsync(e.FeedLines(1));
                await printer.WriteAsync(e.SetStyles(PrintStyle.Bold | PrintStyle.FontB | PrintStyle.DoubleHeight | PrintStyle.DoubleWidth));
                await printer.WriteAsync(e.PrintLine("YOUR LUCKY NUMBERS ARE:"));
                await printer.WriteAsync(e.FeedLines(1));
                await printer.WriteAsync(e.PrintImage(img, true, true, 350, 1100));
                await printer.WriteAsync(e.SetStyles(PrintStyle.None));
                await printer.WriteAsync(e.FeedLines(1));
                await printer.WriteAsync(e.PrintLine("Dont forget to put your name on the back of"));
                await printer.WriteAsync(e.PrintLine("your ticket and place it in the red bin located"));
                await printer.WriteAsync(e.PrintLine("in the bar for your chance to win 1 of 2x $100"));
                await printer.WriteAsync(e.PrintLine("Telegraph vouvhers in our"));
                await printer.WriteAsync(e.PrintLine("end of month draw!"));
                await printer.WriteAsync(e.FeedLines(1));
                await printer.WriteAsync(e.PrintLine("Thanks for visiting The Telegraph Gunning."));
                await printer.WriteAsync(e.PrintLine(DateTime.Now.ToString("dd/MM/yy hh:mm:ss tt")));
                await printer.WriteAsync(e.FeedLines(1));
                await printer.WriteAsync(e.FullCutAfterFeed(1));
            }

            Random rnd = new Random();
            int randomNumber = rnd.Next(1, 5);
            if (randomNumber == 1)
            {
                //tndTicket();
            }
        }

**Comment by xZeroKooLx** [comment]
so i managed to get it working by using System.Collections.Generic.List<T> class to collect the individual byte arrays and then concatenate them into a single array using the System.Linq.Enumerable.SelectMany extension method.

        private async Task PrintNumbersAsync()
        {
            string filePath = "textfile.txt";
            string textFromFile = File.ReadAllText(filePath);
            string[] parts = textFromFile.Split(':');
            string targetAddress = parts[0];
            string targetPort = parts[1];

            if (printer == null) printer = new ImmediateNetworkPrinter(settings: new ImmediateNetworkPrinterSettings() { ConnectionString = $"{targetAddress}:{targetPort}" });

            Bitmap bmp = new Bitmap(listBox1.Width, listBox1.Height);
            listBox1.DrawToBitmap(bmp, new Rectangle(0, 0, bmp.Width, bmp.Height));

            e = new EPSON();
            using (var ms = new MemoryStream())
            {
                bmp.Save(ms, ImageFormat.Png);
                byte[] img = ms.ToArray();

                // Print the image
                var printCommands = new List<byte[]>()
        {
            e.CenterAlign(),
            e.PrintLine(customMsgRTextBox.Text),
            e.FeedLines(1),
            e.SetStyles(PrintStyle.Bold | PrintStyle.FontB | PrintStyle.DoubleHeight | PrintStyle.DoubleWidth),
            e.PrintLine("YOUR LUCKY NUMBERS ARE:"),
            e.FeedLines(1),
            e.PrintImage(img, true, true, 350, 1100),
            e.SetStyles(PrintStyle.None),
            e.FeedLines(1),
            e.PrintLine("Dont forget to put your name on the back of"),
            e.PrintLine("your ticket and place it in the red bin located"),
            e.PrintLine("in the bar for your chance to win 1 of 2x $100"),
            e.PrintLine("Telegraph vouvhers in our"),
            e.PrintLine("end of month draw!"),
            e.FeedLines(1),
            e.PrintLine("Thanks for visiting The Telegraph Gunning."),
            e.PrintLine(DateTime.Now.ToString("dd/MM/yy hh:mm:ss tt")),
            e.FeedLines(1),
            e.FullCutAfterFeed(1)
        };

                var concatenatedCommands = printCommands.SelectMany(c => c).ToArray();
                await printer.WriteAsync(concatenatedCommands);
            }

            Random rnd = new Random();
            int randomNumber = rnd.Next(1, 5);
            if (randomNumber == 1)
            {
                //tndTicket();
            }
        }

This code creates a List<byte[]> to hold each byte array for each print command, and then uses SelectMany to flatten the list into a single array of bytes. Finally, it passes the flattened byte array to printer.WriteAsync() to send the commands to the printer.

**Reply by nedigan** [reply]
Bit unrelated, but where did you find what port number your printer was using to have it work properly as a network printer?
